File: LR4/src/main/java/org/example/types/users/UserType.java
package org.example.types.users;

import org.example.comparator.Comparator;

import java.io.IOException;
import java.io.InputStreamReader;

public interface UserType {

    String typeName();

    Object create() throws InstantiationException;

    Object clone(Object object) throws InstantiationException;

    Object readValue(InputStreamReader in) throws IOException;

    Object parseValue(String ss) throws InstantiationException;

    Comparator getTypeComparator();

    String toString(Object object);
}

File: LR1/src/main/scala/types/users/UserType.scala
package types.users

import comparator.Comparator
import java.io.IOException
import java.io.InputStreamReader

trait UserType {
  def typeName: String

  @throws[InstantiationException]
  def create: AnyRef

  @throws[InstantiationException]
  def clone(`object`: AnyRef): AnyRef

  @throws[IOException]
  def readValue(in: InputStreamReader): Any

  @throws[InstantiationException]
  def parseValue(ss: String): AnyRef

  def getTypeComparator: Comparator

  def toString(`object`: AnyRef): String
}